### PR TITLE
Export and import service principals

### PR DESCRIPTION
**Pull request by ferminho**
Previously, service principals were exported within another processes such as group membership export.

This PR gives service principals the same entity treatment as other users, forking the behavior due to its specifics (no email, generated application Id, specific SCIM API, etc):

- they will be exported in a specific file with a service principal export task
- they will be imported in a specific task, mapping old IDs vs new IDs to be able to, in future PRs, fix service principal ACLs, service principal workspace folders, etc
- they won't be created by other tasks such as group membership import

### CODE REVIEW

**Comment by ferminho** [comment]
Needs https://github.com/OSS-Nextail/databricks-migrate/pull/2 to be merged first. I will rebase afterwards to reduce the clutter in this PR

File: dbclient/WorkspaceClient.py [hunk 1]
```diff
@@ -789,7 +817,7 @@ def import_current_workspace_items(self, artifact_dir='artifacts/'):
     def import_all_workspace_items(self, artifact_dir='artifacts/',
                                    archive_missing=False, num_parallel=4):
         """
-        import all notebooks into a new workspace. Walks the entire artifacts/ directory in parallel, and also
+        import folders and notebooks into a new workspace. Walks the entire artifacts/ directory in parallel, and also
```
Comment (alberto) [inline dbclient/WorkspaceClient.py]:
```suggestion
        import all directories and notebooks into a new workspace. Walks the entire artifacts/ directory in parallel, and also
```

File: dbclient/WorkspaceClient.py [hunk 2]
```diff
@@ -303,6 +303,34 @@ def import_user_home(self, username, local_export_dir):
...
+        checkpoint_dir_set = self._checkpoint_service.get_checkpoint_key_set(
+            wmconstants.WM_EXPORT, wmconstants.WORKSPACE_DIRECTORY_OBJECT)
+        ws_dir_log = self.get_export_dir() + ws_dir_log_file
+        workspace_dir_error_logger = logging_utils.get_error_logger(
+            wmconstants.WM_EXPORT, wmconstants.WORKSPACE_DIRECTORY_OBJECT, self.get_export_dir())
+        if not os.path.exists(ws_dir_log):
+            raise Exception("Run --workspace first to download full log of all directories.")
+        with open(ws_dir_log) as ws_dir_fp:
+            for ws_dir_str in ws_dir_fp:
+                dir_path = json.loads(ws_dir_str).get('path', None).rstrip('\n')
```
Comment (alberto) [inline dbclient/WorkspaceClient.py]: Is this `None` default copied from elsewhere in the codebase? It seems problematic

File: dbclient/ScimClient.py [hunk 3]
```diff
@@ -353,6 +411,37 @@ def assign_user_roles(self, current_user_ids, error_logger, user_log_file='users
                     update_resp = self.patch(f'/preview/scim/v2/Users/{user_id}', patch_roles)
                     logging_utils.log_response_error(error_logger, update_resp)
 
+    def assign_service_principal_roles(self, current_service_principal_ids, error_logger, service_principal_log_file='service_principals.log'):
+        """
+        assign service principal roles that are missing after adding group assignment
+        Note: There is a limitation in the exposed API. If a service principal is assigned a role permission & the permission
+        is granted via a group, we can't distinguish the difference. Only group assignment will be migrated.
```
Comment (alberto) [inline dbclient/ScimClient.py:418]: Not sure I understood this part

**Review by alberto — APPROVED** [review]
LGTM, added a few minor comments

File: dbclient/WorkspaceClient.py (hunk 1 above)
Comment (ferminho) [reply dbclient/WorkspaceClient.py]: thanks!

Comment (ferminho) [reply dbclient/WorkspaceClient.py]: Hm, I don't know why GH says the suggestion is outdated. Will apply manually

Comment (ferminho) [reply dbclient/WorkspaceClient.py]: ah, it's already applied, maybe from rebasing... I guess we did rebase and suggestion around the same time

File: dbclient/ScimClient.py (hunk 3 above)
Comment (ferminho) [reply dbclient/ScimClient.py:418]: It's copy pasted from the same steps done with the users. I didn't fully understand the API limitation thing either, but I copied it to play it safe.
What the user role function does is a diff of the current user roles attribute and the new one, and PATCH the diff. I guess this is about the fact that the "roles" attr might be populated from permissions inherited from groups as well as direct ones and they cannot be differentiated... but the final sentence is still a big question mark for me.

File: dbclient/WorkspaceClient.py (hunk 2 above)
Comment (ferminho) [reply dbclient/WorkspaceClient.py]: Yes, it is. And I agree. Path attr must exist, and I wouldn't do a thorough validation, but better to use ['path'] and get a KeyError than an AttributeError about None and rstrip

Comment (ferminho) [reply dbclient/WorkspaceClient.py]: I understood now about the previous suggestion not being applied, and this one. This modification was done in the previous PR. I started this one from that, and then rebased when we could merge it.
I will do the change in master directly.

**Pull request merged**